clients/command/src/main/java/org/eclipse/hono/client/command/KubernetesContainerInfoProvider.java

```diff
@@ -0,0 +1,230 @@
+        isRunningInKubernetes = System.getenv("KUBERNETES_SERVICE_HOST") != null;
+        podName = isRunningInKubernetes ? System.getenv("HOSTNAME") : null;
+    }
+
+    /**
+     * Gets the KubernetesContainerInfoProvider instance.
+     *
+     * @return The KubernetesContainerInfoProvider.
+     */
+    public static KubernetesContainerInfoProvider getInstance() {
+        return INSTANCE;
+    }
+
+    boolean isRunningInKubernetes() {
+        return isRunningInKubernetes;
+    }
+
+    /**
+     * Gets the Kubernetes pod name if running in Kubernetes.
+     *
+     * @return The pod name or {@code null} if not running in Kubernetes.
+     */
+    public String getPodName() {
+        return podName;
+    }
+
+    /**
+     * Determines the container id if running in a container in Kubernetes.
+     * <p>
+     * First an attempt is made to get the container id by inspecting <code>/proc/self/cgroup</code>
+     * (via {@link CgroupV1KubernetesContainerUtil#getContainerId()}).
+     * If not found there, the container id is queried via the Kubernetes API.
+     * <p>
+     * NOTE: The service account of the application pod must have an RBAC role allowing "get" on the "pods" resource.
+     * If this application is running in a pod with multiple containers, the container that this application is running
+     * in must have an environment variable with the name specified in {@link #KUBERNETES_CONTAINER_NAME_ENV_VAR} set
+     * to the container name.
+     *
+     * @return A future indicating the outcome of the operation.
+     *         The future will be succeeded with the container id or {@code null} if not running in Kubernetes.
+     *         In case of an error determining the container id, the future will be failed with either a
+     *         {@link IllegalStateException} if a precondition for getting the id isn't fulfilled (because of missing
+     *         permissions or because multiple pod containers exist and no KUBERNETES_CONTAINER_NAME env var is set),
+     *         or otherwise a {@link ServerErrorException}.
+     */
+    public Future<String> getContainerId() {
+        if (containerId != null) {
+            return Future.succeededFuture(containerId);
+        }
+        final Context currentContext = Vertx.currentContext();
+        if (currentContext == null) {
+            return Future.failedFuture(new IllegalStateException("not running on vert.x Context"));
+        }
+        if (!isRunningInKubernetes()) {
+            return Future.succeededFuture(null);
+        }
+        final String containerIdViaCgroup1 = CgroupV1KubernetesContainerUtil.getContainerId();
+        if (containerIdViaCgroup1 != null) {
+            containerId = containerIdViaCgroup1;
+            return Future.succeededFuture(containerId);
+        }
+        final Promise<String> containerIdPromise = Promise.promise();
+        if (!containerIdPromiseRef.compareAndSet(null, containerIdPromise)) {
```

Comment:
if this is such an expensive operation and you anticipate this operation to be invoked concurrently/multiple times, wouldn't it make sense to cache the result?

Reply:
The result is already cached via the `containerId` field value.

Reply:
Ah, my bad